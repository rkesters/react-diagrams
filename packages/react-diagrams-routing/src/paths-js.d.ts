declare module 'paths-js/path' {
	export interface Instruction {
		params: number[];
		command: 'default' | 'V' | 'H' | 'A' | 'C' | 'Z';
	}

	export interface Point {
		x: number;
		y: number;
	}

	export interface CurveTo {
		x1: number;
		y1: number;
		x2: number;
		y2: number;
		x: number;
		y: number;
	}

	export interface SmoothCurveTo {
		x2: number;
		y2: number;
		x: number;
		y: number;
	}

	export interface QCurveTo {
		x1: number;
		y1: number;
		x: number;
		y: number;
	}

	export interface Arc {
		rx: number;
		ry: number;
		xrot: number;
		largeArcFlag: number;
		sweepFlag: number;
		x: number;
		y: number;
	}

	export interface Translate {
		dx: number;
		dy: number;
	}

	export interface Rotate {
		angle: number;
		rx: number;
		ry: number;
	}

	export interface Scale {
		sx: number;
		sy: number;
	}

	export interface Shear {

	}

	export interface IPath {
		moveto(point: Point): IPath;
		moveto(x: number, y: number): IPath;
		moveto(x: number | Point, y?: number): IPath;

		lineto(point: Point): IPath;
		lineto(x: number, y: number): IPath;
		lineto(x: number | Point, y?: number): IPath;

		hlineto(point: Point): IPath;
		hlineto(x: number, y: number): IPath;
		hlineto(x: number | Point, y?: number): IPath;

		vlineto(point: Point): IPath;
		vlineto(x: number, y: number): IPath;
		vlineto(x: number | Point, y?: number): IPath;

		closepath(): IPath;

		curveto(curve: CurveTo): IPath;
		curveto(x1: number, y1: number, x2: number, y2: number, x: number): IPath;
		curveto(x1: number | CurveTo, y1?: number, x2?: number, y2?: number, x?: number): IPath;

		smoothcurveto(point: SmoothCurveTo): IPath;
		smoothcurveto(x2: number, y2: number, x: number): IPath;
		smoothcurveto(x2: number | SmoothCurveTo, y2?: number, x?: number): IPath;

		qcurveto(point: QCurveTo): IPath;
		qcurveto(x1: number, y1: number, x: number): IPath;
		qcurveto(x1: number | QCurveTo, y1?: number, x?: number): IPath;

		smoothqcurveto(point: Point): IPath;
		smoothqcurveto(x: number, y: number): IPath;
		smoothqcurveto(x: number | Point, y?: number): IPath;

		arc(point: Arc): IPath;
		arc(rx: number, ry: number, xrot: number, largeArcFlag: number, sweepFlag: number, x: number, y: number): IPath;
		arc(
			rx: number | Arc,
			ry?: number,
			xrot?: number,
			largeArcFlag?: number,
			sweepFlag?: number,
			x?: number,
			y?: number
		): IPath;

		translate(point: Translate): IPath;
		translate(dx: number, dy: number): IPath;
		translate(dx: number | Translate, dy?: number): IPath;

		rotate(point: Rotate): IPath;
		rotate(angle: number, rx: number, ry: number): IPath;
		rotate(angle: number | Rotate, rx: number, ry: number): IPath;

		scale(scaleBy: Scale): IPath;
		scale(sx: number, sy: number): IPath;
		scale(sx: number | Scale, sy?: number): IPath;

		shearX(angle: Scale): IPath;
		shearX(angle: number): IPath;
		shearX(angle: Scale | number): IPath;

		shearY(angle: Scale): IPath;
		shearY(angle: number): IPath;
		shearY(angle: Scale | number): IPath;

		print(): string;
		toString(): string;
		points() : [number, number][];
		instructions() : Instruction[];

		connect(path: IPath): IPath;
	}

	export default function Path(init?: Instruction[]): IPath;
}
