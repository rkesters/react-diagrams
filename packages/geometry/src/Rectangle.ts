import { isNumber } from 'lodash';
import { Point } from './Point';
import { Polygon } from './Polygon';

export class Rectangle extends Polygon {
	constructor(tl: Point, tr: Point, br: Point, bl: Point);
	constructor(position: Point, width: number, height: number);
	constructor(x?: number, y?: number, width?: number, height?: number);

	constructor(a: Point | number = 0, b: Point | number  = 0, c: Point | number  = 0, d: Point | number  = 0) {
		if (Point.isa(a) && Point.isa(b) && Point.isa(c) && Point.isa(d)) {
			super([a, b, c, d]);
			return;
		} else if (Point.isa(a) && isNumber(b) && isNumber(c) && isNumber(d)) {
			super([a, new Point(a.x + b, a.y), new Point(a.x + b, a.y + c), new Point(a.x, a.y + c)]);
			return;
		} else if (isNumber(a) && isNumber(b) && isNumber(c) && isNumber(d)) {
			super(Rectangle.pointsFromBounds(a, b, c, d));
			return;
		}

		throw new Error(`Rectangle call with invaild arguements: ${arguments.length}`);
	}

	static pointsFromBounds(x: number, y: number, width: number, height: number): Point[] {
		return [new Point(x, y), new Point(x + width, y), new Point(x + width, y + height), new Point(x, y + height)];
	}

	updateDimensions(x: number, y: number, width: number, height: number) {
		this.points = Rectangle.pointsFromBounds(x, y, width, height);
	}

	setPoints(points: Point[]) {
		if (points.length !== 4) {
			throw 'Rectangles must always have 4 points';
		}
		super.setPoints(points);
	}

	containsPoint(point: Point) {
		const tl = this.getTopLeft();
		const br = this.getBottomRight();

		return point.x >= tl.x && point.x <= br.x && point.y >= tl.y && point.y <= br.y;
	}

	getWidth(): number {
		return Math.sqrt(
			Math.pow(this.getTopLeft().x - this.getTopRight().x, 2) + Math.pow(this.getTopLeft().y - this.getTopRight().y, 2)
		);
	}

	getHeight(): number {
		return Math.sqrt(
			Math.pow(this.getBottomLeft().x - this.getTopLeft().x, 2) +
				Math.pow(this.getBottomLeft().y - this.getTopLeft().y, 2)
		);
	}

	getTopMiddle(): Point {
		return Point.middlePoint(this.getTopLeft(), this.getTopRight());
	}

	getBottomMiddle(): Point {
		return Point.middlePoint(this.getBottomLeft(), this.getBottomRight());
	}

	getLeftMiddle(): Point {
		return Point.middlePoint(this.getBottomLeft(), this.getTopLeft());
	}

	getRightMiddle(): Point {
		return Point.middlePoint(this.getBottomRight(), this.getTopRight());
	}

	getTopLeft(): Point {
		return this.points[0];
	}

	getTopRight(): Point {
		return this.points[1];
	}

	getBottomRight(): Point {
		return this.points[2];
	}

	getBottomLeft(): Point {
		return this.points[3];
	}
}
