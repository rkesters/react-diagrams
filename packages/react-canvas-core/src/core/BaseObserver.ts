import { has, isNil } from 'lodash';
import { Toolkit } from '../Toolkit';

import debug from 'debug';

const dbg = debug('reactDiagrams:canvasCore:BaseObserver');
export interface BaseEvent {
	firing: boolean;
	stopPropagation: () => any;
}

export interface BaseEventProxy extends BaseEvent {
	function: string;
}

/**
 * Listeners are always in the form of an object that contains methods that take events
 */
export type BaseListener = {
	/**
	 * Generic event that fires before a specific event was fired
	 */
	eventWillFire?: (event: BaseEvent & { function: string }) => void;

	/**
	 * Generic event that fires after a specific event was fired (even if it was consumed)
	 */
	eventDidFire?: (event: BaseEvent & { function: string }) => void;
	/**
	 * Type for other events that will fire
	 */
	[key: string]: (event: BaseEvent) => any;
};

export interface ListenerHandle {
	/**
	 * Used to degister the listener
	 */
	deregister: () => any;
	/**
	 * Original ID of the listener
	 */
	id: string;

	/**
	 * Original Listener
	 */
	listener: BaseListener;
}

export function isaListenerHandle(value: unknown): value is ListenerHandle {
	return has(value, 'id') && has(value, 'deregister') && has(value, 'listener');
}

/**
 * Base observer pattern class for working with listeners
 */
export class BaseObserver<L extends BaseListener = BaseListener> {
	protected listeners: { [id: string]: L };

	constructor() {
		this.listeners = {};
	}

	private fireEventInternal(fire: boolean, k: keyof L, event: BaseEvent) {
		dbg(`fireEventInternal ${k} ${fire}`);
		this.iterateListeners((listener) => {
			dbg(`iterateListeners ${k} ${fire}`);
			// returning false here will instruct itteration to stop
			if (!fire && !event.firing) {
				dbg(`iterateListeners ${k} will not fire ${fire} ${event.firing}`);

				return false;
			}

			dbg(`iterateListeners ${k} ${isNil(listener[k])}`);
			// fire selected listener
			if (listener[k]) {
				listener[k](event as BaseEvent);
			}
		});
	}

	fireEvent<K extends keyof L>(event: Partial<Parameters<L[K]>[0]>, k: keyof L) {
		dbg(`fireEvent ${k}`);
		event = {
			firing: true,
			stopPropagation: () => {
				event.firing = false;
			},
			...event
		};

		// fire pre
		this.fireEventInternal(true, 'eventWillFire', {
			...event,
			function: k
		} as BaseEventProxy);

		dbg(`fireEvent ${k} Main`);
		// fire main event
		this.fireEventInternal(false, k, event as BaseEvent);
		dbg(`fireEvent ${k} post`);
		// fire post
		this.fireEventInternal(true, 'eventDidFire', {
			...event,
			function: k
		} as BaseEventProxy);
	}

	iterateListeners(cb: (listener: L) => any) {
		dbg(`iterateListeners (${Object.keys(this.listeners ?? {}).join(',')})`);
		for (let id in this.listeners) {
			const res = cb(this.listeners[id]);
			// cancel itteration on false
			if (res === false) {
				return;
			}
		}
	}

	getListenerHandle(listener: L): ListenerHandle {
		for (let id in this.listeners) {
			if (this.listeners[id] === listener) {
				return {
					id: id,
					listener: listener,
					deregister: () => {
						delete this.listeners[id];
					}
				};
			}
		}
	}

	registerListener(listener: L): ListenerHandle {
		const id = Toolkit.UID();
		this.listeners[id] = listener;
		const modelID = (this as any).getID ? (this as any).getID() : 'no id';
		dbg(`registerListener on ${modelID} : (${Object.keys(this.listeners ?? {}).join(',')}) (${Object.keys(listener ?? {}).join(',')})`);
		return {
			id: id,
			listener: listener,
			deregister: () => {
			dbg(`deregister on ${modelID} (${Object.keys(this.listeners ?? {}).join(',')}) (${Object.keys(listener ?? {}).join(',')})`);
				delete this.listeners[id];
			}
		};
	}

	deregisterListener(listener: L | ListenerHandle) {
		if (isaListenerHandle(listener)) {
			(listener as ListenerHandle).deregister();
			return true;
		}
		const handle = this.getListenerHandle(listener);
		if (handle) {
			dbg(`deregisterListener (${handle.id})`);

			handle.deregister();
			return true;
		}
		return false;
	}
}
