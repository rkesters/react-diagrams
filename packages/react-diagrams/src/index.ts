import {
	DefaultDiagramState,
	DiagramEngine,
	DiagramModel,
	LinkLayerFactory,
	NodeLayerFactory
} from '@projectstorm/react-diagrams-core';
import {
	DefaultLabelFactory,
	DefaultLinkFactory,
	DefaultNodeFactory,
	DefaultPortFactory
} from '@projectstorm/react-diagrams-defaults';
import { PathFindingLinkFactory } from '@projectstorm/react-diagrams-routing';
import { SelectionBoxLayerFactory, CanvasEngineOptions } from '@projectstorm/react-canvas-core';

export * from '@projectstorm/react-diagrams-core';
export * from '@projectstorm/react-diagrams-defaults';
export * from '@projectstorm/react-diagrams-routing';

/**
 * Construct an engine with the defaults installed
 */
export default (model: DiagramModel, options: CanvasEngineOptions = {}): DiagramEngine => {
	const engine = new DiagramEngine(model,options);

	// register model factories
	engine.getLayerFactories().registerFactory(new NodeLayerFactory(engine));
	engine.getLayerFactories().registerFactory(new LinkLayerFactory(engine));
	engine.getLayerFactories().registerFactory(new SelectionBoxLayerFactory(engine));

	engine.getLabelFactories().registerFactory(new DefaultLabelFactory(engine));
	engine.getNodeFactories().registerFactory(new DefaultNodeFactory(engine)); // i cant figure out why
	engine.getLinkFactories().registerFactory(new DefaultLinkFactory(engine));
	engine.getLinkFactories().registerFactory(new PathFindingLinkFactory(engine));
	engine.getPortFactories().registerFactory(new DefaultPortFactory(engine));

	// register the default interaction behaviours
	engine.getStateMachine().pushState(new DefaultDiagramState(engine));
	return engine;
};
