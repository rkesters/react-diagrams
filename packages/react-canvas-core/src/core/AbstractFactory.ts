import { CanvasEngine } from '../CanvasEngine';
import { FactoryBank } from './FactoryBank';

/**
 * Base factory for all the different types of entities.
 * Gets registered with the engine, and is used to generate models
 */
export abstract class AbstractFactory<E extends CanvasEngine = CanvasEngine> {

	protected bank: FactoryBank | undefined | null;

	constructor(protected type: string, protected engine: E) {
		this.type = type;
	}

	setDiagramEngine(engine: E): void {
		this.engine = engine;
	}

	setFactoryBank(bank: FactoryBank<any> | null): void {
		this.bank = bank;
	}

	getType(): string {
		return this.type;
	}
}
