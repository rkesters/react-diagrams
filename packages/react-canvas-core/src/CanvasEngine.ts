import { debounce } from 'lodash';
import { CanvasModel } from './entities/canvas/CanvasModel';
import { FactoryBank } from './core/FactoryBank';
import { AbstractReactFactory } from './core/AbstractReactFactory';
import { LayerModel } from './entities/layer/LayerModel';
import { BaseListener, BaseObserver } from './core/BaseObserver';
import { MouseEvent } from 'react';
import { BaseModel } from './core-models/BaseModel';
import { Point } from '@projectstorm/geometry';
import { ActionEventBus } from './core-actions/ActionEventBus';
import { ZoomCanvasAction } from './actions/ZoomCanvasAction';
import { DeleteItemsAction } from './actions/DeleteItemsAction';
import { StateMachine } from './core-state/StateMachine';
export interface CanvasEngineListener extends BaseListener {
	canvasReady(): void;

	repaintCanvas(): void;

	rendered(): void;
}

/**
 * Defines the CanvasEngine options
 */
export interface CanvasEngineOptions {
	registerDefaultDeleteItemsAction?: boolean;
	registerDefaultZoomCanvasAction?: boolean;
	/**
	 * Defines the debounce wait time in milliseconds if > 0
	 */
	repaintDebounceMs?: number;
}

export class CanvasEngine<
	L extends CanvasEngineListener = CanvasEngineListener,
	M extends CanvasModel = CanvasModel
> extends BaseObserver<L> {
	protected model: M ;
	protected layerFactories: FactoryBank<AbstractReactFactory<LayerModel>>;
	protected canvas: HTMLDivElement | undefined;
	protected eventBus: ActionEventBus;
	protected stateMachine: StateMachine;
	protected options: CanvasEngineOptions;

	constructor(model: M, options: CanvasEngineOptions = {}) {
		super();
		this.model = model;
		this.eventBus = new ActionEventBus(this);
		this.stateMachine = new StateMachine(this);
		this.layerFactories = new FactoryBank();
		this.registerFactoryBank(this.layerFactories);

		/**
		 * Overrides the standard options with the possible given options
		 */
		this.options = {
			registerDefaultDeleteItemsAction: true,
			registerDefaultZoomCanvasAction: true,
			repaintDebounceMs: 0,
			...options
		};
		if (this.options.registerDefaultZoomCanvasAction === true) {
			this.eventBus.registerAction(new ZoomCanvasAction({}, this));
		}
		if (this.options.registerDefaultDeleteItemsAction === true) {
			this.eventBus.registerAction(new DeleteItemsAction({}, this));
		}
	}

	getStateMachine() {
		return this.stateMachine;
	}

	getRelativeMousePoint(event: { clientX: number; clientY: number }): Point {
		const point = this.getRelativePoint(event.clientX, event.clientY);
		return new Point(
			(point.x - (this.model?.getOffsetX() ?? 0)) / ((this.model?.getZoomLevel() ?? 1) / 100.0),
			(point.y - (this.model?.getOffsetY() ?? 0)) / ((this.model?.getZoomLevel() ?? 1) / 100.0)
		);
	}

	getRelativePoint(x:number, y: number): Point {
		const canvasRect = this.canvas?.getBoundingClientRect();
		if (!canvasRect) {
			throw new Error(`Can't get relative point because canvas has not been defined.`)
		}
		return new Point(x - canvasRect.left, y - canvasRect.top);
	}

	registerFactoryBank(factory: FactoryBank<any>) {
		factory.registerListener({
			factoryAdded: (event) => {
				event.factory.setDiagramEngine(this);
			},
			factoryRemoved: (event) => {
				event.factory.setDiagramEngine(null);
			}, eventWillFire: ():void => {} , eventDidFire: ():void => {}
		});
	}

	getActionEventBus() {
		return this.eventBus;
	}

	getLayerFactories() {
		return this.layerFactories;
	}

	getFactoryForLayer<F extends AbstractReactFactory<LayerModel>>(layer: LayerModel | string | undefined) {
		if (!layer) { return;}
		if (typeof layer === 'string') {
			return this.layerFactories.getFactory(layer);
		}
		return this.layerFactories.getFactory(layer.getType());
	}

	setModel(model: M) {
		this.model = model;
		if (this.canvas) {
			requestAnimationFrame(() => {
				this.repaintCanvas();
			});
		}
	}

	getModel(): M {
		return this.model;
	}

	repaintCanvas(promise: true): Promise<any>;
	repaintCanvas(): void;
	repaintCanvas(promise?: true): Promise<any> | void {
		const repaintDebounceMs  = this.options.repaintDebounceMs ?? 0;

		/**
		 * The actual repaint function
		 */
		const repaint = () => {
			this.iterateListeners((listener) => {
				if (listener.repaintCanvas) {
					listener.repaintCanvas();
				}
			});
		};

		// if the `repaintDebounceMs` option is > 0, then apply the debounce
		let repaintFn = repaint;

		if (repaintDebounceMs > 0) {
			repaintFn = debounce(repaint, repaintDebounceMs);
		}

		if (promise) {
			return new Promise<void>((resolve) => {
				const l = this.registerListener({
					rendered: () => {
						resolve();
						l.deregister();
					}
				} as L);
				repaintFn();
			});
		}

		repaintFn();
	}

	setCanvas(canvas?: HTMLDivElement | null) {
		if (this.canvas !== canvas) {
			this.canvas = canvas ?? undefined;
			if (canvas) {
				this.fireEvent(undefined, 'canvasReady');
			}
		}
	}

	getCanvas() {
		return this.canvas;
	}

	getMouseElement(event: MouseEvent): BaseModel<any> | null | undefined  {
		return null;
	}

	zoomToFit() {
		if (!this.canvas) {
			throw new Error(`Can not zoom to fit because canvas is not defined`)
		}
		const xFactor = this.canvas.clientWidth / this.canvas.scrollWidth;
		const yFactor = this.canvas.clientHeight / this.canvas.scrollHeight;
		const zoomFactor = xFactor < yFactor ? xFactor : yFactor;

		this.model?.setZoomLevel((this.model?.getZoomLevel() ?? 1) * zoomFactor);
		this.model?.setOffset(0, 0);
		this.repaintCanvas();
	}
}
