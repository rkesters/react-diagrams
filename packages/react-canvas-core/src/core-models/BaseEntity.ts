import * as _ from 'lodash';
import { CanvasEngine } from '../CanvasEngine';
import { BaseEvent, BaseListener, BaseObserver, ParameterFirst } from '../core/BaseObserver';
import { Toolkit } from '../Toolkit';
import { BaseModel } from './BaseModel';

export interface BaseEntityEvent<T extends BaseEntity = BaseEntity> extends BaseEvent {
	entity?: T;
}

export interface LockChangedEvent<T extends BaseEntity = BaseEntity> extends BaseEntityEvent<T>  {
	locked: boolean;
}
export interface BaseEntityListener extends BaseListener {
	lockChanged(event: LockChangedEvent ): void;
}

export type BaseEntityType = 'node' | 'link' | 'port' | 'point';

export interface BaseEntityOptions {
	id?: string;
	locked?: boolean;
}

export type BaseEntityGenerics = {
	LISTENER: BaseEntityListener;
	OPTIONS: BaseEntityOptions;
};

export interface DeserializeEvent<S = ReturnType<BaseEntity['serialize']>> {
	engine: CanvasEngine;
	data: S;
	registerModel(model: BaseModel): void;
	getModel<T extends BaseModel>(id: string | null): Promise<T | null>;
}

type ff<T extends BaseEntityGenerics = BaseEntityGenerics> = ParameterFirst<T['LISTENER']['lockChanged'], BaseEvent>
const FF: ff = {locked: false, entity: {} as BaseEntity} ;
export class BaseEntity<T extends BaseEntityGenerics = BaseEntityGenerics> extends BaseObserver<T['LISTENER']> {
	protected options: T['OPTIONS'];

	protected id: string;
	constructor(options: T['OPTIONS'] = {}) {
		super();
		this.id = options.id ?? Toolkit.UID();
		this.options = {
			id: this.id,
			...options
		};
	}

	getOptions(): T['OPTIONS'] {
		return this.options;
	}

	getID(): string {
		return this.id;
	}

	doClone(lookupTable: { [s: string]: any } = {}, clone: any): void {
		/*noop*/
	}

	clone(lookupTable: { [s: string]: any } = {}) {
		// try and use an existing clone first
		if (lookupTable[this.id]) {
			return lookupTable[this.id];
		}
		let clone = _.cloneDeep(this);
		clone.options = {
			...this.options,
			id: Toolkit.UID()
		};
		clone.clearListeners();
		lookupTable[this.id] = clone;

		this.doClone(lookupTable, clone);
		return clone;
	}

	clearListeners() {
		this.listeners = {};
	}

	deserialize(event: DeserializeEvent<ReturnType<BaseEntity['serialize']>>) {
		this.options.id = event.data.id;
		this.options.locked = event.data.locked;
	}

	serialize() {
		return {
			id: this.options.id,
			locked: this.options.locked
		};
	}

	// fireEvent<Event extends keyof Methods<L, B>,>( k: keyof T['LISTENER'], event: L) {
	// 	super.fireEvent(
	// 		{
	// 			entity: this,
	// 			...event
	// 		},
	// 		k
	// 	);
	// }

	public isLocked(): boolean {
		return this.options.locked ?? false;
	}

	public setLocked(locked: boolean = true) {
		this.options.locked = locked;

		this.fireEvent(
			{ locked: true },
			'lockChanged'
		);
	}
}
