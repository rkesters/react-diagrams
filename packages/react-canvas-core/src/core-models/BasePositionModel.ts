import { BaseModel, BaseModelGenerics, BaseModelListener, BaseModelOptions } from './BaseModel';
import { BaseEntityEvent, DeserializeEvent } from './BaseEntity';
import { Point, Rectangle } from '@projectstorm/geometry';
import { ModelGeometryInterface } from '../core/ModelGeometryInterface';

export interface BasePositionModelListener extends BaseModelListener {
	positionChanged(event: BaseEntityEvent<BasePositionModel>): void;
}

export interface BasePositionModelOptions extends BaseModelOptions {
	position?: Point;
}

export interface BasePositionModelGenerics extends BaseModelGenerics {
	LISTENER: BasePositionModelListener;
	OPTIONS: BasePositionModelOptions;
}

export class BasePositionModel<G extends BasePositionModelGenerics = BasePositionModelGenerics>
	extends BaseModel<G>
	implements ModelGeometryInterface
{
	protected position: Point;

	constructor(options: G['OPTIONS']) {
		super(options);
		this.position = options.position || new Point(0, 0);
	}

	setPosition(point: Point): void;
	setPosition(x: number, y: number): void;
	setPosition(x: Point | number , y?: number | undefined ) {
		this.position = Point.isa(x) ? x : new Point(x,y as number);
		this.fireEvent({}, 'positionChanged');
	}

	getBoundingBox(): Rectangle {
		return new Rectangle(this.position, 0, 0);
	}

	deserialize(event: DeserializeEvent<ReturnType<BasePositionModel['serialize']>>) {
		super.deserialize(event);
		this.position = new Point(event.data.x, event.data.y);
	}

	serialize() {
		return {
			...super.serialize(),
			x: this.position.x,
			y: this.position.y
		};
	}

	getPosition(): Point {
		return this.position;
	}

	getX(): number {
		return this.position.x;
	}

	getY() {
		return this.position.y;
	}
}
