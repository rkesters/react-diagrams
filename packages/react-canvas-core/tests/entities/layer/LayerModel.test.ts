import { map } from 'lodash';
import { CanvasEngine, CanvasEngineListener } from '../../../src/CanvasEngine';
import { BaseModel, BaseModelGenerics } from '../../../src/core-models/BaseModel';
import { GenerateModelEvent, AbstractModelFactory } from '../../../src/core/AbstractModelFactory';
import { AbstractReactFactory, GenerateWidgetEvent } from '../../../src/core/AbstractReactFactory';
import { FactoryBank, FactoryBankListener } from '../../../src/core/FactoryBank';
import { CanvasModel, CanvasModelGenerics } from '../../../src/entities/canvas/CanvasModel';
import { LayerModel, LayerModelGenerics, LayerModelOptions } from '../../../src/entities/layer/LayerModel';

class TestLayerFactory extends AbstractModelFactory<
	BaseModel<BaseModelGenerics>,
	CanvasEngine<CanvasEngineListener, CanvasModel<CanvasModelGenerics>>
> {
	constructor() {
		super('test');
	}
	generateModel(event: GenerateModelEvent): BaseModel<BaseModelGenerics> {
		return new BaseModel({});
	}
}

class TestLayer extends LayerModel {

	#childModelFactoryBank = new FactoryBank<
		AbstractModelFactory<
			BaseModel<BaseModelGenerics>,
			CanvasEngine<CanvasEngineListener, CanvasModel<CanvasModelGenerics>>
		>,
		FactoryBankListener<
			AbstractModelFactory<
				BaseModel<BaseModelGenerics>,
				CanvasEngine<CanvasEngineListener, CanvasModel<CanvasModelGenerics>>
			>
		>
	>();

	constructor(options: LayerModelOptions= {}) {
		super({...options, type: 'test' });

		this.#childModelFactoryBank.registerFactory(new TestLayerFactory());
	}
	getChildModelFactoryBank(
		engine: CanvasEngine<CanvasEngineListener, CanvasModel<CanvasModelGenerics>>
	): FactoryBank<
		AbstractModelFactory<
			BaseModel<BaseModelGenerics>,
			CanvasEngine<CanvasEngineListener, CanvasModel<CanvasModelGenerics>>
		>,
		FactoryBankListener<
			AbstractModelFactory<
				BaseModel<BaseModelGenerics>,
				CanvasEngine<CanvasEngineListener, CanvasModel<CanvasModelGenerics>>
			>
		>
	> {
		return this.#childModelFactoryBank;
	}
}

describe('LayerModel', () => {
	test('Contructor', () => {
		let model = new TestLayer();

		expect(model.getOptions()).toMatchSnapshot({ id: expect.any(String) });

		 model = new TestLayer({isSvg: true, transformed: true, type: 'bad'});

		expect(model.getOptions()).toMatchSnapshot({ id: expect.any(String) });
	});

	test('serialize', () => {
		const model = new TestLayer();

		let serial = model.serialize();

		expect(serial).toMatchSnapshot({ id: expect.any(String) });

		serial.id = 'dog';

		model.deserialize({ data: serial } as any);

		expect(model.serialize()).toMatchSnapshot();

		model.addModel(new BaseModel({ type: 'test' }));
		model.addModel(new BaseModel({ type: 'test' }));

		serial = model.serialize();
		model.deserialize({ data: serial } as any);

		expect(Object.keys(model.getModels()).length).toEqual(2);
		expect(Object.keys(model.getModels())).toEqual(Object.keys(serial.models));
	});

	test('getters/setters', () => {
		const model = new TestLayer();

		expect(model.isRepaintEnabled()).toEqual(true);
		model.allowRepaint(false);
		expect(model.isRepaintEnabled()).toEqual(false);
		model.allowRepaint();
		expect(model.isRepaintEnabled()).toEqual(true);
		model.allowRepaint(false);
		expect(model.isRepaintEnabled()).toEqual(false);
		model.allowRepaint(true);
		expect(model.isRepaintEnabled()).toEqual(true);

		model.addModel(new BaseModel({ type: 'test', extras: { name: 'unselected' } }));
		model.addModel(new BaseModel({ type: 'test', selected: true, extras: { name: 'selected' } }));

		const selected = model.getSelectionEntities();
		const ms = map( model.getModels(), (value) => value)

		expect(selected).toEqual(ms);

		expect(model.getModel(ms[0].getID())).toEqual(ms[0]);

		expect(model.removeModel(ms[0].getID())).toBeTruthy()
		expect(model.getModel(ms[0].getID())).toBeUndefined();

		expect(model.removeModel(ms[1])).toBeTruthy()
		expect(model.getModel(ms[1].getID())).toBeUndefined();
		expect(model.removeModel(ms[1].getID())).toBeFalsy();

		const parent = new CanvasModel();
		model.setParent(parent)
		const removeListener = {entityRemoved: jest.fn()};
		model.registerListener(removeListener);
		model.remove();
		expect(removeListener.entityRemoved).toHaveBeenCalled();


	});
});
