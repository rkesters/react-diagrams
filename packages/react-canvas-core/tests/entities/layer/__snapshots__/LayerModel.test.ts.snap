// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LayerModel Contructor 1`] = `
Object {
  "id": Any<String>,
  "type": "test",
}
`;

exports[`LayerModel Contructor 2`] = `
Object {
  "id": Any<String>,
  "isSvg": true,
  "transformed": true,
  "type": "test",
}
`;

exports[`LayerModel serialize 1`] = `
Object {
  "extras": undefined,
  "id": Any<String>,
  "isSvg": undefined,
  "locked": undefined,
  "models": Object {},
  "selected": undefined,
  "transformed": undefined,
  "type": "test",
}
`;

exports[`LayerModel serialize 2`] = `
Object {
  "extras": undefined,
  "id": "dog",
  "isSvg": false,
  "locked": undefined,
  "models": Object {},
  "selected": undefined,
  "transformed": false,
  "type": "test",
}
`;
